File: code-coaches/medium/camel-to-snake/camel_to_snake.rb
def to_snake_case(camel_case_string)
  snake_case_string = ""

  camel_case_string.each_char.with_index do |char, i|
    if i != 0
      previous_char = camel_case_string[i - 1]
      if char.upcase == char && previous_char =~ /[a-zA-Z]/ || char =~ /[a-zA-Z]/ && previous_char =~ /\d/
        snake_case_string += "_"
      end
    end

    snake_case_string += char.downcase
  end

  snake_case_string
end

def main
  camel_case_string = gets.chomp
  puts to_snake_case(camel_case_string)
end

if __FILE__ == $0
  main
end

File: code-coaches/medium/camel-to-snake/CamelToSnake.swift
func toSnakeCase(camelCaseString: String) -> String {
    var snakeCaseString = ""

    for (i, char) in camelCaseString.enumerated() {
        if i != 0 {
            let previousChar = camelCaseString[camelCaseString.index(camelCaseString.startIndex, offsetBy: i - 1)]

            let isPreviousCharAlpha = (previousChar >= "a" && previousChar <= "z") || (previousChar >= "A" && previousChar <= "Z")
            let isCharAlpha = (char >= "a" && char <= "z") || (char >= "A" && char <= "Z")
            let isCharUppercased = char >= "A" && char <= "Z"
            let isCharNumber = char >= "0" && char <= "9"

            if (isCharUppercased && isPreviousCharAlpha) || (isCharAlpha && isCharNumber) {
                snakeCaseString += "_"
            }
        }
        snakeCaseString += String(char).lowercased()
    }

    return snakeCaseString
}

func main() {
    if let camelCaseString = readLine() {
        print(toSnakeCase(camelCaseString: camelCaseString))
    }
}

main()